src/boot/efi/efi-string.c

```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
 DEFINE_PARSE_NUMBER(char, parse_number8);
 DEFINE_PARSE_NUMBER(char16_t, parse_number16);
 
+static void format_number(
+                char16_t buf[static SCRATCH_BUF_SIZE],
+                size_t min_digits,
+                bool hex,
+                bool hex_prefix,
+                bool lower,
+                bool is_signed,
+                uint64_t u) {
+
+        size_t n = 0;
+        int base = hex ? 16 : 10;
+        char16_t tmp[SCRATCH_BUF_SIZE];
+        const char16_t * const digits = lower ? u"0123456789abcdef" : u"0123456789ABCDEF";
+
+        if (is_signed && (int64_t) u < 0) {
+                buf[0] = '-';
+                buf++;
+
+                int rem = -((int64_t) u % base);
+                u = (int64_t) u / -base;
+                tmp[n++] = digits[rem];
+        }
+
+        while (u > 0 || n == 0) {
+                int rem = u % base;
+                u /= base;
+                tmp[n++] = digits[rem];
+        }
+
+        assert(n + min_digits + 5 < SCRATCH_BUF_SIZE);
+
+        if (hex_prefix) {
+                buf[0] = '0';
+                buf[1] = lower ? 'x' : 'X';
+                buf += 2;
+        }
+
+        while (min_digits > n) {
+                buf[0] = '0';
+                buf++;
+                min_digits--;
+        }
+
+        while (n > 0) {
+                *buf = tmp[--n];
+                buf++;
+        }
+
+        *buf = '\0';
+}
+
+static const char *parse_format_specifier(
+                EFI_STATUS status,
+                const char *format,
+                va_list *ap,
+                char16_t tmp[SCRATCH_BUF_SIZE],
+                bool *ret_pad_zero,
+                int *ret_precision,
+                int *ret_field_width,
+                const char **ret_str8,
+                const char16_t **ret_str16,
+                const wchar_t **ret_wstr) {
+
+        assert(ap);
+        assert(ret_pad_zero);
+        assert(ret_precision);
+        assert(ret_field_width);
+        assert(ret_str8);
+        assert(ret_str16);
+        assert(ret_wstr);
+
+        bool alternative_form = false, have_field_width = false, long_arg = false;
+        size_t field_length = sizeof(int);
+
+        for (;; format++)
+                switch (*format) {
+                case '0':
+                        *ret_pad_zero = true;
+                        break;
+                case '.':
+                        have_field_width = true;
+                        break;
+                case '#':
+                        alternative_form = true;
+                        break;
+
+                case '*':
+                case '1' ... '9': {
+                        uint64_t u;
+
+                        if (*format == '*')
+                                u = va_arg(*ap, int);
+                        else {
+                                if (!parse_number8(format, &u, &format) || u > INT_MAX)
+                                        assert_not_reached();
```

Comment:
Even if the format is checked by the compiler with `_printf_` attribute, I'd like to return the function with `-EINVAL` and entirely refuse to formatting the remaining string.

Reply:
I would argue that any invalid format string goes to the same category as assert on internal function parameters. If we are fine with "don't call this function with a NULL pointer", it's only logical to apply the same for "don't call this with invalid/unsupported format string". All these are internal functions and aren't supposed to be exposed to outside users, so imho asserting should be safe and keeps things simpler.

Assert also makes it much more likely that an invalid/unsupported format does not get unnoticed instead of being silently swallowed somewhere.